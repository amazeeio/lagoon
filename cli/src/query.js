// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { config } from './config/';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.lagoon-token');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const apiUrl = R.prop('api', config) || 'https://api.amazee.io/graphql';

  const {
    hostname, path: pathname, port: urlPort, protocol,
  } = url.parse(
    apiUrl,
  );

  if (hostname == null) {
    throw new Error(
      'API URL configured under the "api" key in .lagoon.yml doesn\'t contain a valid hostname.',
    );
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port:
      urlPort === null
        ? R.propOr(443, protocol)(protocolPorts)
        : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  try {
    return await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always('Could not connect to API.'),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }
}

export default runGQLQuery;
