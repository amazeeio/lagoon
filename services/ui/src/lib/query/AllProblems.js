import gql from 'graphql-tag';

export default gql`
  {
    allGroups {
      name
      projects {
        name
        id
        environments(type: PRODUCTION) {
          name
          id
          problems {
            id
            identifier
            severity
            source
          }
        }
      }
    }
  }
`;
