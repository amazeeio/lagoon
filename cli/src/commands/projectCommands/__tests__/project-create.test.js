// @flow

import R from 'ramda';
import { runGQLQuery } from '../../../query';

import {
  CUSTOMER,
  OPENSHIFT,
  allOptionsSpecified,
  commandOptions,
  getAllowedCustomersAndOpenshifts,
  createProject,
} from '../create';

jest.mock('../../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const options = R.mapObjIndexed(
  (value, key) =>
    // If the key is CUSTOMER or OPENSHIFT...
    new RegExp(`${CUSTOMER}|${OPENSHIFT}`).test(key)
      ? 0 // ...return a number to satisfy the Flow type...
      : value, // ...otherwise just return the value
  commandOptions,
);

describe('allOptionsSpecified', () => {
  it('should return true when all options are specified', () => {
    const returnVal = allOptionsSpecified(options);
    expect(returnVal).toBe(true);
  });

  it('should return true when all options and more are specified', () => {
    const returnVal = allOptionsSpecified({
      ...options,
      anotherOptionKey: 'anotherOptionValue',
    });
    expect(returnVal).toBe(true);
  });

  it('should return false if not all options are specified', () => {
    const returnVal = allOptionsSpecified({
      ...R.dissoc(R.head(R.keys(commandOptions)), commandOptions),
    });
    expect(returnVal).toBe(false);
  });
});

describe('getAllowedCustomersAndOpenshifts', () => {
  it('should return all customers and all openshifts', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allCustomers: [
            {
              value: 1,
              name: 'credentialtest-customer1',
            },
            {
              value: 2,
              name: 'credentialtest-customer2',
            },
          ],
          allOpenshifts: [
            {
              value: 1,
              name: 'credentialtest-openshift',
            },
          ],
        },
      }),
    );

    const cerr = jest.fn();

    const returnVal = await getAllowedCustomersAndOpenshifts(cerr);

    expect(R.prop('allCustomers', returnVal)).toHaveLength(2);
    expect(R.prop('allOpenshifts', returnVal)).toHaveLength(1);
    expect(returnVal).toMatchSnapshot();
  });
});

describe('createProject', () => {
  it('should display table after successful project creation', async () => {
    _mock(runGQLQuery)
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            allCustomers: [
              {
                value: 1,
                name: 'credentialtest-customer1',
              },
              {
                value: 2,
                name: 'credentialtest-customer2',
              },
            ],
            allOpenshifts: [
              {
                value: 1,
                name: 'credentialtest-openshift',
              },
            ],
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            addProject: {
              id: 1,
              name: 'test-project',
              customer: {
                name: 'credentialtest-customer1',
              },
              git_url: 'ssh://git@192.168.99.1:2222/git/project1.git',
              active_systems_deploy: 'lagoon_openshiftBuildDeploy',
              active_systems_remove: 'lagoon_openshiftRemove',
              branches: true,
              pullrequests: true,
              production_environment: null,
              openshift: 1,
              created: '2018-03-05 10:26:22',
            },
          },
        }),
      );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await createProject({
      clog,
      cerr,
      options: {
        customer: 1,
        name: 'test-project',
        git_url: 'ssh://git@192.168.99.1:2222/git/project1.git',
        openshift: 1,
        branches: 'true',
        pullrequests: 'true',
        production_environment: 'null',
      },
    });

    expect(returnVal).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it("should display error, if GraphQL sends error messages the first time it's called", async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        errors: [{ message: 'something something error' }],
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await createProject({ clog, cerr, options: {} });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it("should display error, if GraphQL sends error messages the second time it's called", async () => {
    _mock(runGQLQuery)
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            allCustomers: [
              {
                value: 1,
                name: 'credentialtest-customer1',
              },
              {
                value: 2,
                name: 'credentialtest-customer2',
              },
            ],
            allOpenshifts: [
              {
                value: 1,
                name: 'credentialtest-openshift',
              },
            ],
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          errors: [{ message: 'something something error 2' }],
        }),
      );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await createProject({
      clog,
      cerr,
      options: {
        customer: 1,
        name: 'test-project',
        git_url: 'ssh://git@192.168.99.1:2222/git/project1.git',
        openshift: 1,
        branches: 'true',
        pullrequests: 'true',
        production_environment: 'null',
      },
    });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL response contains zero customers', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        data: { allCustomers: [], allOpenshifts: [] },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await createProject({ clog, cerr, options: {} });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL response contains zero openshifts but one customer', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allCustomers: [
            {
              value: 1,
              name: 'credentialtest-customer1',
            },
          ],
          allOpenshifts: [],
        },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await createProject({ clog, cerr, options: {} });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });
});
