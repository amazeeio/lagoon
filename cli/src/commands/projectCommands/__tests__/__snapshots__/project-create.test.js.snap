// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProject should display error, if GraphQL response contains zero customers 1`] = `
Array [
  Array [
    "[31mNo authorized customers found![39m",
  ],
]
`;

exports[`createProject should display error, if GraphQL response contains zero openshifts but one customer 1`] = `
Array [
  Array [
    "[31mNo authorized openshifts found![39m",
  ],
]
`;

exports[`createProject should display error, if GraphQL sends error messages the first time it's called 1`] = `
Array [
  Array [
    "[31mOops! The lagoon API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"something something error\\"[39m
[31m}[39m",
  ],
]
`;

exports[`createProject should display error, if GraphQL sends error messages the second time it's called 1`] = `
Array [
  Array [
    "[31mOops! The lagoon API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"something something error 2\\"[39m
[31m}[39m",
  ],
]
`;

exports[`createProject should display table after successful project creation 1`] = `
Array [
  Array [
    "[34m![39m Using \\"customer\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"name\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"git_url\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"openshift\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"branches\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"pullrequests\\" option from arguments or config",
  ],
  Array [
    "[34m![39m Using \\"production_environment\\" option from arguments or config",
  ],
  Array [
    "[32mProject \\"test-project\\" created successfully:[39m",
  ],
  Array [
    "╔═══════════════════════╤══════════════════════════════════════════════╗
║ Project Name          │ test-project                                 ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Customer              │ credentialtest-customer1                     ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Git URL               │ ssh://git@192.168.99.1:2222/git/project1.git ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy                  ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove                       ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Branches              │ true                                         ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Pull Requests         │ true                                         ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Openshift             │                                              ║
╟───────────────────────┼──────────────────────────────────────────────╢
║ Created               │ 2018-03-05 10:26:22                          ║
╚═══════════════════════╧══════════════════════════════════════════════╝
",
  ],
]
`;

exports[`getAllowedCustomersAndOpenshifts should return all customers and all openshifts 1`] = `
Object {
  "allCustomers": Array [
    Object {
      "name": "credentialtest-customer1",
      "value": 1,
    },
    Object {
      "name": "credentialtest-customer2",
      "value": 2,
    },
  ],
  "allOpenshifts": Array [
    Object {
      "name": "credentialtest-openshift",
      "value": 1,
    },
  ],
  "errors": undefined,
}
`;
